astropy/nddata/nduncertainty.py

```
@@ -30,155 +84,187 @@ class MissingDataAssociationException(Exception):
+       propagation if their ``parent_nddata`` attribute is a reference to the
+       `NDData` object. This is needed since many kinds of propagation need the
+       actual data besides the uncertainty.
+
+    6. If a ``unit`` is implicit (``data`` had a unit) or explicitly passed
+       to the ``__init__`` one should be sure that this unit is identical or
+       convertable to the unit of the parent. Otherwise uncertainty propagation
+       should fail.
     """
 
-    __metaclass__ = abc.ABCMeta
-
-    # Indicates whether the class supports the propagation of correlated
-    # uncertainties
-    supports_correlated = False
+    def __init__(self, array=None, unit=None, copy=True):
+        if isinstance(array, NDUncertainty):
+            # Given an NDUncertainty class or subclass check that the type
+            # is the same.
+            if array.uncertainty_type != self.uncertainty_type:
+                raise IncompatibleUncertaintiesException
+            # Check if two units are given and take the explicit one then.
+            if (unit is not None and array.unit is not None and
+                    unit != array.unit):
+                # TODO : Clarify it (see NDData.init for same problem)?
+                log.info("Overwriting Uncertainty's current "
+                         "unit with specified unit")
+            elif array.unit is not None:
+                unit = array.unit
+            array = array.array
 
-    @property
-    def parent_nddata(self):
-        if self._parent_nddata is None:
-            message = "uncertainty is not associated with an NDData object"
-            raise MissingDataAssociationException(message)
+        elif isinstance(array, Quantity):
+            # Check if two units are given and take the explicit one then.
+            if (unit is not None and array.unit is not None and
+                    unit != array.unit):
+                # TODO : Clarify it (see NDData.init for same problem)?
```

You silently accepted it in #4270 (I think there is still the TODO). That was if given two units (one from data, one explicitly) should raise an error or raise a warning.